airflow/lineage/__init__.py

```diff
@@ -133,13 +133,20 @@ def wrapper(self, context, *args, **kwargs):
 
             # Remove auto and task_ids
             self.inlets = [i for i in self.inlets if not isinstance(i, str)]
-            _inlets = self.xcom_pull(context, task_ids=task_ids, dag_id=self.dag_id, key=PIPELINE_OUTLETS)
-
-            # re-instantiate the obtained inlets
-            # xcom_pull returns a list of items for each given task_id
-            _inlets = [item for item in itertools.chain.from_iterable(_inlets)]
-
-            self.inlets.extend(_inlets)
+            # We manually create a session here since xcom_pull returns a LazyXComAccess iterator.
+            # If we do not pass a session a new session will be created, however that session will not be
+            # properly closed and will remain open. After we are done iterating we can safely close this
+            # session.
+            with create_session() as session:
+                _inlets = self.xcom_pull(
+                    context, task_ids=task_ids, dag_id=self.dag_id, key=PIPELINE_OUTLETS, session=session
+                )
+
+                # re-instantiate the obtained inlets
+                # xcom_pull returns a list of items for each given task_id
+                _inlets = [item for item in itertools.chain.from_iterable(_inlets)]
+
+                self.inlets.extend(_inlets)
```

Comment:
```suggestion
                self.inlets.extend(i for i in itertools.chain.from_iterable(_inlets))
```

don’t need the intermediate list, I think